### PR TITLE
Storage API

### PR DESCRIPTION
**Pull request by dvoytenko**

### CODE REVIEW

File: src/service/storage-impl.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+ * @return {!Storage}
+ */
+export function installStorageService(window) {
+  return getService(window, 'storage', () => {
+    const viewer = installViewerService(window);
+    const overrideStorage = parseInt(viewer.getParam('storage'), 10);
+    const binding = overrideStorage ?
+        new ViewerStorageBinding(window) :
+        new LocalStorageBinding(window);
+    return new Storage(window, viewer, binding).start_();
```
Comment (ericfs) [inline src/service/storage-impl.js]: Won't this return undefined?

File: src/service/storage-impl.js [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+ */
+export class ViewerStorageBinding {
+
+  /**
+   * @param {!Window} win
+   * @param {!Viewer} viewer
+   */
+  constructor(win, viewer) {
+    /** @const {!Window} */
+    this.win = win;
```
Comment (ericfs) [inline src/service/storage-impl.js]: not used

File: src/service/storage-impl.js (hunk 1 above)
Comment (dvoytenko) [inline src/service/storage-impl.js]: Fixed. Didn't actually test this code yet :)

File: src/service/storage-impl.js (hunk 2 above)
Comment (dvoytenko) [inline src/service/storage-impl.js]: Removed.

**Comment by dvoytenko** [comment]
@ericfs requesting the formal review.

File: src/service/storage-impl.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+   * @override
+   */
+  set(name, value) {
+    return this.saveStore_(store => {
+      store[name] = value;
+    });
+  }
+
+  /**
+   * Saves the specified property. Returns the promise that's resolved when
```
Comment (ericfs) [inline src/service/storage-impl.js]: s/Saves/Removes/

File: src/service/storage-impl.js [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+/**
+ * A binding provides the specific implementation of storage technology.
+ * @interface
+ */
+class StorageBindingDef {
+
+  /**
+   * Returns the promise that yields the store blob for the specified origin.
+   * @param {string} unusedOrigin
```
Comment (ericfs) [inline src/service/storage-impl.js]: Not unused in implementations. Do you always name interface parameters unused?

Comment (dvoytenko) [inline src/service/storage-impl.js]: Yes, very unfortunately. The linter doesn't understand that this is an interface :(

File: src/service/storage-impl.js (hunk 3 above)
Comment (dvoytenko) [inline src/service/storage-impl.js]: Done

File: src/service/storage-impl.js [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    return new Promise(resolve => {
+      resolve(this.win.localStorage.getItem(this.getKey_(origin)) || {});
+    });
+  }
+
+  /** @override */
+  saveStore(origin, store) {
+    return new Promise(resolve => {
+      this.win.localStorage.setItem(this.getKey_(origin), store);
+      resolve();
```
Comment (ericfs) [inline src/service/storage-impl.js]: Why not use Promise.resolve() here?

File: test/functional/test-storage.js [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+
+  it('should initialize with experiment', () => {
+    expect(viewerBroadcastHandler).to.exist;
+  });
+
+  it('should not initialize without experiment', () => {
+    viewerBroadcastHandler = undefined;
+    windowApi.document.cookie = '';
+    new Storage(windowApi, viewer, binding).start_();
+    expect(storage.viewerBroadcastHandler).to.not.exist;
```
Comment (ericfs) [inline test/functional/test-storage.js]: is the use of storage correct?

File: src/service/storage-impl.js (hunk 5 above)
Comment (dvoytenko) [inline src/service/storage-impl.js]: This gives me automatic rejection if `setItem` method fails without needing to use try/catch directly.

File: test/functional/test-storage.js (hunk 6 above)
Comment (dvoytenko) [inline test/functional/test-storage.js]: Good catch. Not correct. Fixed.

**Comment by ericfs** [comment]
LGTM

**Comment by dvoytenko** [comment]
I did some additional research today and need a little time to process it. I will hold off merging this PR in the meantime. @ericfs please hold off viewer work on this for a little bit as well.

**Comment by ericfs** [comment]
Ok, let me know if you're updating the PR.

File: src/service/storage-impl.js [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+  /**
+   * @return {!Promise<!JSONObject>}
+   * @private
+   */
+  getStore_() {
+    if (this.store_) {
+      return Promise.resolve(this.store_);
+    }
+    return this.whenStarted_.then(() => {
```
Comment (jridgewell) [inline src/service/storage-impl.js]: Why not save this as a `storePromise` to avoid the extra allocations above?

Comment (dvoytenko) [inline src/service/storage-impl.js]: Good idea. Done.

**Comment by dvoytenko** [comment]
Assigning to @cramforce for additional review.

File: src/service/storage-impl.js [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+    /** @const @private {boolean} */
+    this.isExperimentOn_ = isExperimentOn(this.win, EXPERIMENT);
+
+    /** @private @const {!Promise} */
+    this.whenStarted_ = this.isExperimentOn_ ?
+        Promise.resolve() :
+        Promise.reject(`Enable experiment ${EXPERIMENT}`);
+
+    /** @private {!Promise<!JSONObject>} */
```
Comment (jridgewell) [inline src/service/storage-impl.js]: Can you use `!` with a null promise?

Comment (dvoytenko) [inline src/service/storage-impl.js]: Nope. Fixed. (Where are you closure compiler?)

File: src/amp-core-service.js [hunk 9]
```diff
@@ -38,4 +39,5 @@ export function installCoreServices(window) {
   installResourcesService(window);
   installStandardActions(window);
   installFramerateService(window);
+  installStorageService(window);
```
Comment (cramforce) [inline src/amp-core-service.js]: Lets move this into `amp-analytics` for now. I'd like to really keep core lean.

File: src/service/storage-impl.js [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+  /**
+   * Saves the value of the specified property. Returns the promise that's
+   * resolved when the operation completes.
+   * @param {string} name
+   * @param {*} value
+   * @return {!Promise}
+   * @override
+   */
+  set(name, value) {
+    return this.saveStore_(store => {
```
Comment (cramforce) [inline src/service/storage-impl.js]: Lets enforce values here to be boolean.

File: src/service/storage-impl.js [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+    return this.storePromise_;
+  }
+
+  /**
+   * @param {function(!JSONObject)} mutator
+   * @return {!Promise}
+   * @private
+   */
+  saveStore_(mutator) {
+    return this.getStore_().then(store => {
```
Comment (cramforce) [inline src/service/storage-impl.js]: Lets enforce a maximum count. E.g. 8 entries for now.

File: src/amp-core-service.js (hunk 9 above)
Comment (dvoytenko) [inline src/amp-core-service.js]: Done

File: src/service/storage-impl.js (hunk 10 above)
Comment (dvoytenko) [inline src/service/storage-impl.js]: Done

File: src/service/storage-impl.js (hunk 11 above)
Comment (dvoytenko) [inline src/service/storage-impl.js]: Done. I had to switch the format of the storage to accommodate this.

**Comment by dvoytenko** [comment]
@cramforce PTAL. The format of the storage changed considerably to allow limits. In general, though, it looks like a good idea.

File: src/service/storage-impl.js [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+   */
+  constructor(obj, opt_maxValues) {
+    /** @const {!JSONObject} */
+    this.obj = obj;
+
+    /** @private @const {number} */
+    this.maxValues_ = opt_maxValues || MAX_VALUES_PER_ORIGIN;
+
+    /** @private @const {!Array<!JSONObject>} */
+    this.values_ = obj['vv'] || [];
```
Comment (jridgewell) [inline src/service/storage-impl.js]: Why not use an `Object<string, JSONObject>`? That'll avoid the `O(n)` lookup times.

Comment (dvoytenko) [inline src/service/storage-impl.js]: Possible of course. The set is very small - forced to be under 8 elements. It simplifies search and removal of old values. And it also seemed a bit better as a format. But I can definitely switch to a map if you feel that format would improve.

Comment (jridgewell) [inline src/service/storage-impl.js]: Oh, I see why you're doing it now (purging old values). Different comments incoming. 😉

File: src/service/storage-impl.js [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+      this.values_[index]['v'] = value;
+      this.values_[index]['t'] = timer.now();
+    } else {
+      this.values_.push({'k': name, 'v': value, 't': timer.now()});
+    }
+
+    // Purge old values.
+    if (this.values_.length > this.maxValues_) {
+      let minIndex = 0;
+      for (let i = 1; i < this.values_.length; i++) {
```
Comment (jridgewell) [inline src/service/storage-impl.js]: Starting at `1` will prevent us from checking if `0` can be expired.

Comment (dvoytenko) [inline src/service/storage-impl.js]: `minIndex` starts with 0.

Comment (jridgewell) [inline src/service/storage-impl.js]: IE. imagine `0` and `1` are both expired. We'll always splice out `1` with this code, where `0` should be the one to go.

Comment (dvoytenko) [inline src/service/storage-impl.js]: This code changed, although, no, it'd still would have gone out with `0` :)

**Comment by dvoytenko** [comment]
Switched the new version to use hash format.

**Comment by dvoytenko** [comment]
@cramforce PTAL.

File: src/service/storage-impl.js [hunk 14]
```diff
@@ -0,0 +1,381 @@
...
+  /**
+   * Saves the value of the specified property. Returns the promise that's
+   * resolved when the operation completes.
+   * @param {string} name
+   * @param {*} value
+   * @return {!Promise}
+   * @override
+   */
+  set(name, value) {
+    assert(typeof value == 'boolean', 'Only boolean values accepted');
```
Comment (cramforce) [inline src/service/storage-impl.js]: We need to make sure to not provide a "read all keys" API. That could be used to put values into keys :)

**Comment by cramforce** [comment]
LGTM

**Comment by dvoytenko** [comment]
@ericfs this is merged. I don't think viewer bindings have really changed at all.

**Pull request merged**